contracts/libs/0.8.x/BytecodeStorage.sol

```diff
@@ -0,0 +1,257 @@
+// SPDX-License-Identifier: LGPL-3.0-only
+// Created By: Art Blocks Inc.
+
+pragma solidity ^0.8.0;
+
+/**
+ * @title Art Blocks Script Storage Library
+ * @notice Utilize contract bytecode as persistant storage for large chunks of script string data.
+ *
+ * @author Art Blocks Inc.
+ * @author Modified from 0xSequence (https://github.com/0xsequence/sstore2/blob/master/contracts/SSTORE2.sol)
+ * @author Modified from Solmate (https://github.com/transmissions11/solmate/blob/main/src/utils/SSTORE2.sol)
+ *
+ * @dev Compared to the above two rerferenced libraries, this contracts-as-storage implementation makes a few
+ *      notably different design decisions:
+ *      - uses the `string` data type for input/output on reads, rather than speaking in bytes directly
+ *      - exposes "delete" functionality, allowing no-longer-used storage to be purged from chain state
+ *      - stores the "writer" address (library user) in the deployed contract bytes, which is useful for both:
+ *         a) providing necessary information for safe deletion; and
+ *         b) allowing this to be introspected on-chain
+ *      Also, given that much of this library is written in assembly, this library makes use of a slightly
+ *      different convention (when compared to the rest of the Art Blocks smart contract repo) around
+ *      pre-defining return values in order to simplify need to directly memory manage these return values.
+ */
+library BytecodeStorage {
+    //---------------------------------------------------------------------------------------------------------------//
+    // Offset Amount | Offset Aggregate | Description                                                                //
+    //---------------------------------------------------------------------------------------------------------------//
+    // 53            | 53               | allow contract to be `selfdestruct`-able via gated-cleanup-logic           //
+    // 20            | 73               | reserve 20 bytes for storing deploying-contract's address                  //
+    //---------------------------------------------------------------------------------------------------------------//
+    // define the offset for where the "logic bytes" end, and the "data bytes" begin
+    uint256 internal constant DATA_OFFSET = 73;
```

Comment:
Unfortunately, I don't think that this is possible without writing some sort of custom semantic analysis tool or assembly-generator based on config files or something like that which is out of scope and a bit scary heh – but yes this would definitely be very aggressively caught by current test coverage if we messed this offset up.

@ryley-o any thoughts on additional readily-usable ways to enforce this?

Reply:
I've updated comments in this commit: https://github.com/ArtBlocks/artblocks-contracts/pull/299/commits/fd2bbf836e8834b5526829dbd2afecccf0863b3d

Cc @jakerockland @shantanubala